### PR TITLE
Product Query: Add support for filtering by attributes within the block

### LINKED ISSUE #7322: Product Query [3]: [Settings] `Filter by Attributes`
Implement a filter setting within the inspector controls to filter products by their attributes.

Design TBD. CC @vivialice

### PR DESCRIPTION
**Pull request by sunyatasattva**
This PR introduces the capability for merchants to filter products within the Product Query block according to their attributes. As the final design is a WIP, this PR implements an autocomplete approach that allows for the selection of the available taxonomies and terms.

Fixes #7322

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

https://user-images.githubusercontent.com/1847066/203838891-5a8a1eec-1ad1-4a52-9733-9ee3391bf3a7.mov

#### User Facing Testing

1. Go to the editor and add a “Product Query” block.
2. Ensure a new “Product Attributes” input is accessible via the “Advanced Filters” panel in the inspector controls.
3. Ensure that the input autocompletes with all the available terms for your store in the format `${Taxonomy}: ${Term}`.
4. Ensure that when selecting the attributes, the products filter in the editor.
5. Ensure that when you remove an attribute, the editor preview also gets updated.
6. Save the post.
7. Ensure that the block displays correctly in the front-end.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Changelog

> Product Query: Add support for filtering by attributes within the block

### CODE REVIEW

File: src/BlockTypes/ProductQuery.php [hunk 1]
```diff
@@ -412,10 +447,12 @@ private function get_queries_by_applied_filters() {
...
-		$on_sale_enabled = isset( $query['__woocommerceOnSale'] ) && true === $query['__woocommerceOnSale'];
+		$query            = $parsed_block['attrs']['query'];
+		$on_sale_enabled  = isset( $query['__woocommerceOnSale'] ) && true === $query['__woocommerceOnSale'];
+		$attributes_query = isset( $query['__woocommerceAttributes'] ) ? $this->get_product_attributes_query( $query['__woocommerceAttributes'] ) : array();
 
 		return array(
+			'attributes'   => $attributes_query,
 			'on_sale'      => ( $on_sale_enabled ? $this->get_on_sale_products_query() : array() ),
 			'stock_status' => isset( $query['__woocommerceStockStatus'] ) ? $this->get_stock_status_query( $query['__woocommerceStockStatus'] ) : array(),
 		);
```
Comment (dinhtungdu) [inline src/BlockTypes/ProductQuery.php:440]: We're using three different ways to prepare data for the three attributes, should we make it more consistent here?

File: assets/js/blocks/product-query/useProductAttributes.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			hasLoadedAttributes.current = true;
+			setIsLoadingAttributes( false );
+		}
+
+		fetchTerms();
+
+		return () => {
+			hasLoadedAttributes.current = true;
+		};
+	}, [ STORE_ATTRIBUTES, isLoadingAttributes, shouldLoadAttributes ] );
```
Comment (dinhtungdu) [inline assets/js/blocks/product-query/useProductAttributes.ts:59]: Great implementation 🚀 TIL, thank you!

File: src/BlockTypes/ProductQuery.php [hunk 3]
```diff
@@ -115,8 +115,9 @@ public function update_query( $pre_render, $parsed_block ) {
 	public function update_rest_query( $args, $request ) {
 		$on_sale_query = $request->get_param( '__woocommerceOnSale' ) !== 'true' ? array() : $this->get_on_sale_products_query();
 		$orderby_query = $this->get_custom_orderby_query( $request->get_param( 'orderby' ) );
+		$tax_query     = $this->get_product_attributes_query( $request->get_param( '__woocommerceAttributes' ) );
 
-		return array_merge( $args, $on_sale_query, $orderby_query );
+		return array_merge( $args, $on_sale_query, $orderby_query, $tax_query );
```
Comment (dinhtungdu) [inline src/BlockTypes/ProductQuery.php:127]: Reviewing this make me realize that I can do the same for #7682 as long as we only have one `meta_query` ( for `get_product_attributes_query()` ) and one `tax_query` (for `get_product_attributes_query()`). And we should keep this simple implementation until we can not use `array_merge` anymore, in the other worlds, until we need to merge more than one `meta_query`/`tax_query`. What do you think?

**Review by dinhtungdu — APPROVED** [review]
The PR is testing well and the code is very solid! Thanks for your excellent PR!

I notice a flashing content issue when loading the attributes, but because the design is WIP, I just note it here as the optional comment for awareness. Please check the screencast for the detail.

<details>
<summary>Screencast</summary>


https://user-images.githubusercontent.com/5423135/204519266-1ad56e90-9d41-4b7f-a331-a0eea1779b28.mov



</details>

File: src/BlockTypes/ProductQuery.php (hunk 1 above)
Comment (sunyatasattva) [reply src/BlockTypes/ProductQuery.php:440]: Yes, good point! Next time I touch this I'll refactor to make it consistent.

File: src/BlockTypes/ProductQuery.php (hunk 3 above)
Comment (sunyatasattva) [reply src/BlockTypes/ProductQuery.php:127]: Totally agree: the simpler, the better.

**Comment by gigitux** [comment]
I found an issue with this PR. On the editor side, when I add the Product Query block, the request fails and returns an error because we don't check that the parameter `__woocommerceAttributes` is defined before invoking `get_product_attributes_query`. (https://github.com/woocommerce/woocommerce-blocks/pull/7743/files#diff-efb947e345af5e501dfaa2980b055b3b31c308b31e8db23049b6b48fecee33d8R125).

This is the error:

```
[29-Nov-2022 15:53:44 UTC] PHP Warning:  array_values() expects parameter 1 to be array, null given in /Users/gigitux/work/woocommerce-gutenberg-products-block/src/BlockTypes/ProductQuery.php on line 313
[29-Nov-2022 15:55:20 UTC] PHP Warning:  array_reduce() expects parameter 1 to be array, null given in /Users/gigitux/work/woocommerce-gutenberg-products-block/src/BlockTypes/ProductQuery.php on line 309
```

**Comment by sunyatasattva** [comment]
Thank you! Will fix it tomorrow!

**Pull request merged**